### ISSUE TITLE
Dynamic styles inside `&` scope are applied to every component

### ISSUE THREAD

**Issue by nd0ut** [styled-components/styled-components#4194]
## Environment

- Node: 18.12.1
- npm: 8.19.2
- styled-components@6.1.0

## Reproduction

https://stackblitz.com/edit/stackblitz-starters-eplxdo?file=src%2FApp.js

## Steps to reproduce

```jsx
const Link = styled.a`
& {
  color: ${({ color }) => color};
}
`;

<Link color="black" href="#">black</Link>
<Link color="red" href="#">red</Link>
```

## Expected Behavior

Each link have its own color.

## Actual Behavior

![image](https://github.com/styled-components/styled-components/assets/670959/7073ec92-0be5-48d1-95a5-7522a3f327fd)

Both links are red.

**Comment by jeshio** [comment]
I have the same problem

**Comment by bricejar** [comment]
Hi, same problem here.

**Comment by chuanqisun** [comment]
Minimum repro https://stackblitz.com/edit/stackblitz-starters-q4mqvr?file=src%2FApp.js

```jsx
import React from 'react';
import './style.css';
import styled from 'styled-components';

export default function App() {
  return (
    <div>
      <div>
        <GoodDiv>Expect red background, actual red background</GoodDiv>
      </div>
      <div>
        <BadDiv>Expect red background, actual no background</BadDiv>
      </div>
    </div>
  );
}

const GoodDiv = styled.div`
&:first-child {
  background-color: red
}
`;

const BadDiv = styled.div`
:first-child {
  background-color: red
}
`;

```

**Comment by chuanqisun** [comment]
Related issue: #4177

**Comment by horneber** [comment]
> Minimum repro https://stackblitz.com/edit/stackblitz-starters-q4mqvr?file=src%2FApp.js
> 
> ```js
> import React from 'react';
> import styled from 'styled-components';
> 
> export default function App() {
>   return (
>     <div>
>       <div>
>         <GoodDiv>Expect red background, actual red background</GoodDiv>
>       </div>
>       <div>
>         <BadDiv>Expect red background, actual no background</BadDiv>
>       </div>
>     </div>
>   );
> }
> 
> const GoodDiv = styled.div`
> &:first-child {
>   background-color: red
> }
> `;
> 
> const BadDiv = styled.div`
> :first-child {
>   background-color: red
> }
> `;
> ```
> 
> For additional context, prior to 6.0.9, the `&` was ignored as well. So both of the cases above would fail.

Is this really the same issue though? OPs case shows css attached to the static class that should not have any css attached to it, according to https://styled-components.com/docs/faqs#why-do-my-dom-nodes-have-two-classes

If this were the same issue, I'd expect the example quoted here to actually have red background for both styled `div`s.

**Comment by chuanqisun** [comment]
> ```js
> color: ${({ color }) => color};
> ```

Thanks @horneber for calling that out. I misunderstood the bug. And I could not repro the OP's issue. If anyone else can repro it, please chime it.

**Comment by horneber** [comment]
I haven't tried OP's minimal example, but a larger codebase I'm migrating from v5 to v6 shows this exact behavior under very similar circumstances.

I.e. CSS that's using `&` gets attached to static classes and is applied to every component using the same HTML tag.

I'll try out OP's minimal case tomorrow.

**Comment by horneber** [comment]
Reproduced with OP's minimal example.

**Comment by lukaswiklund** [comment]
I agree with @domarmstrong. This issue alters a lot of our styles, preventing us from upgrading to v6.

**Comment by quantizor** [comment]
This should be fixed with 6.1.6

**Comment by quantizor** [comment]
Closing — this was fixed in v6.1.5 via commit b6fe70d6, which tightened the self-referencing `&` replacement logic in the stylis plugin.
